### PR TITLE
Migrate RemovePodsHavingTooManyRestarts to plugin

### PR DESCRIPTION
**Pull request by BinacsLee**

### CODE REVIEW

File: pkg/framework/plugins/removepodshavingtoomanyrestarts/toomanyrestarts.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+func validateCanEvict(pod *v1.Pod, tooManyRestartsArgs *componentconfig.RemovePodsHavingTooManyRestartsArgs) error {
+	var errs []error
+
+	restarts := calcContainerRestartsFromStatuses(pod.Status.ContainerStatuses)
+	if tooManyRestartsArgs.IncludingInitContainers {
+		restarts += calcContainerRestartsFromStatuses(pod.Status.InitContainerStatuses)
+	}
+
+	if restarts < tooManyRestartsArgs.PodRestartThreshold {
+		errs = append(errs, fmt.Errorf("pod's restarts %v is excluded", restarts))
```
Comment (ingvagabund) [inline pkg/framework/plugins/removepodshavingtoomanyrestarts/toomanyrestarts.go:126]: There's at most a single error. No need to store the error into an array:
```
var err error
...
err = fmt.Errorf("pod's restarts %v is excluded", restarts)
...
return err
```
will do as well.

Comment (ingvagabund) [reply pkg/framework/plugins/removepodshavingtoomanyrestarts/toomanyrestarts.go:126]: s/"pod's restarts %v is excluded"/"number of container restarts (%v) not exceeding the threshold"

File: pkg/framework/plugins/removepodshavingtoomanyrestarts/toomanyrestarts_test.go [hunk 2]
```diff
@@ -121,100 +120,115 @@ func TestRemovePodsHavingTooManyRestarts(t *testing.T) {
...
 		maxNoOfPodsToEvictPerNamespace *uint
+		nodeFit                        bool
 	}{
 		{
 			description:             "All pods have total restarts under threshold, no pod evictions",
-			strategy:                createStrategy(true, true, 10000, false),
+			args:                    createRemovePodsHavingTooManyRestartsAgrs(10000, true),
 			nodes:                   []*v1.Node{node1},
 			expectedEvictedPodCount: 0,
+			nodeFit:                 false,
```
Comment (ingvagabund) [inline pkg/framework/plugins/removepodshavingtoomanyrestarts/toomanyrestarts_test.go:134]: This line can be omitted, `false` is the default value for `nodeFit`. The same for other lines.

File: test/e2e/e2e_toomanyrestarts_test.go [hunk 3]
```diff
@@ -146,33 +156,33 @@ func TestTooManyRestarts(t *testing.T) {
...
+					SharedInformerFactoryImpl:     sharedInformerFactory,
+					GetPodsAssignedToNodeFuncImpl: getPodsAssignedToNode,
+				})
+			if err != nil {
+				t.Fatalf("Unable to initialize the plugin: %v", err)
+			}
+
+			// Run RemovePodsHavingTooManyRestarts strategy
+			t.Log("Running RemovePodsHavingTooManyRestarts strategy")
+			plugin.(framework.DeschedulePlugin).Deschedule(ctx, nodes)
```
Comment (ingvagabund) [inline test/e2e/e2e_toomanyrestarts_test.go:183]: s/nodes/workerNodes

File: test/e2e/e2e_toomanyrestarts_test.go [hunk 4]
```diff
@@ -30,24 +30,27 @@ import (
...
+	clientSet, sharedInformerFactory, _, getPodsAssignedToNode, stopCh := initializeClient(t)
 	defer close(stopCh)
 
 	nodeList, err := clientSet.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
 	if err != nil {
 		t.Errorf("Error listing node with %v", err)
 	}
 
-	nodes, workerNodes := splitNodesAndWorkerNodes(nodeList.Items)
+	nodes, _ := splitNodesAndWorkerNodes(nodeList.Items)
```
Comment (ingvagabund) [inline test/e2e/e2e_toomanyrestarts_test.go:53]: It is still important to run the plugin over worker nodes only. Some of the control plane pod containers may get restarted too many times and evicted by the descheduler. Changing the assumption about the master nodes being protected against eviction and making the debugging step more complicated.

**Review by ingvagabund — CHANGES_REQUESTED** [review]

**Comment by ingvagabund** [comment]
@BinacsLee thank You for helping us with the migration to plugins. It's great to see improvements in the code.

**Comment by BinacsLee** [comment]
hi @ingvagabund , thanks for your review! I have fixed the code, PTAL. :)

**Comment by ingvagabund** [comment]
@BinacsLee thank You for addressing the comment. Would you please rebase the PR as well?

**Comment by BinacsLee** [comment]
@ingvagabund sure, rebase done. PTAL

**Comment by BinacsLee** [comment]
sorry, will fix the test soon

File: pkg/apis/componentconfig/types_pluginargs.go [hunk 5]
```diff
@@ -57,3 +57,13 @@ type RemovePodsViolatingNodeAffinityArgs struct {
 	LabelSelector    *metav1.LabelSelector
 	NodeAffinityType []string
 }
+
```
Comment (ingvagabund) [inline pkg/apis/componentconfig/types_pluginargs.go:60]: Missing `// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object` line

Comment (BinacsLee) [reply pkg/apis/componentconfig/types_pluginargs.go:60]: Thanks a lot! `pull-descheduler-verify-master` can be passed now

**Pull request merged**